### PR TITLE
[Hinty] Type hinting: prepare tests & setup Mypy

### PR DESCRIPTION
**Pull request by gpotter2**
This PR provides the tools, scripts and tests required to start "Project Hinty" = adding type hinting into Scapy.

- provides UTscapy util to use `pyannotate` to generate type hinting
- add Mypy tests to check regressions
- relocate test configs
- some various fixes that were failing on my test machine
- two files were used (processed) as a trial: `main.py` and `__init__.py`. Because `http2.py` had also been done previously, it was also included.

### CODE REVIEW

File: scapy/__init__.py [hunk 1]
```diff
@@ -38,6 +39,9 @@ def _version_from_git_describe():
 
         >>> _version_from_git_describe()
         '2.3.2.dev346'
+
+    :raises CalledProcessError: if git is unavailable
+    :return: the scapy latest tag
```
Comment (guedou) [inline scapy/__init__.py]: s/scapy/Scapy/

File: scapy/__init__.py [hunk 2]
```diff
@@ -62,6 +66,11 @@ def _version_from_git_describe():
 
 
 def _version():
+    # () -> str
+    """Returns the Scapy version from multiple methods
+
+    :return: the scapy version
```
Comment (guedou) [inline scapy/__init__.py]: s/scapy/Scapy/

File: scapy/__init__.py [hunk 3]
```diff
@@ -62,6 +66,11 @@ def _version_from_git_describe():
 
 
 def _version():
+    # () -> str
```
Comment (guedou) [inline scapy/__init__.py]: Should it start by `type:` ?

File: scapy/main.py [hunk 4]
```diff
@@ -52,6 +52,7 @@
 
 
 def _probe_config_file(cf):
+    # str -> str
```
Comment (guedou) [inline scapy/main.py]: Should it start by `type:` ?

File: scapy/tools/UTscapy.py [hunk 5]
```diff
@@ -696,7 +703,7 @@ def usage():
 -T\t\t: if -t is used with *, remove a specific file (can be used many times)
 -l\t\t: generate local .js and .css files
 -F\t\t: expand only failed tests
--b\t\t: stop at first failed campaign
+-b\t\t: don't stop at the first failed campaign
```
Comment (guedou) [inline scapy/tools/UTscapy.py]: Why did you change that?

Comment (gpotter2) [reply scapy/tools/UTscapy.py]: The default was already True.

File: scapy/main.py (hunk 4 above)
Comment (gpotter2) [reply scapy/main.py]: Indeed

File: .mypy/mypy_check.py [hunk 6]
```diff
@@ -0,0 +1,32 @@
+# This file is part of Scapy
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply .mypy/mypy_check.py]: I expect this to break the configs, but let's try it anyway

File: .mypy/mypy_check.py [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+# Load files
+
+with io.open("./.mypy/mypy_enabled.txt") as fd:
+    FILES = [l.strip() for l in fd.readlines() if l.strip() and l[0] != "#"]
+
+if not FILES:
+    print("No files specified. Arborting")
+    sys.exit(0)
+
+# Generate Mypy arguments
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply .mypy/mypy_check.py]: They seem to be quite okay with both:
image

**Comment by gpotter2** [comment]
This is going to be quite some work. After having done `main.py` and `__init__.py` as a starter to test this PR, we'll probably have to do most files one by one (one per PR).

**Comment by gpotter2** [comment]
Another note: `typing` isn't installed on Python 2.7 (but is on Python 3.5+).
This adds a dependency to Scapy for Python 2.7. Scapy will still be able to be ran "out-of-the-box" for Python 3.5+.

Should we then choose to remove `run_scapy_py2.bat` ?

File: scapy/main.py [hunk 8]
```diff
@@ -31,9 +31,11 @@
 from scapy.themes import DefaultTheme, BlackAndWhite, apply_ipython_style
 from scapy.consts import WINDOWS
 
+from typing import cast, Any, Dict, List, Optional, Union
```
Comment (guedou) [inline scapy/main.py]: The `six` module does not handle `typing`. I think that we can handle it in `scapy/compat.py`. If the typing module is installed, we return the real type. If not, we return a fake one.

File: setup.py [hunk 9]
```diff
@@ -41,6 +41,10 @@ def get_long_description():
             'UTscapy = scapy.tools.UTscapy:main'
         ]
     },
+    # This is already installed by default on Python >= 3.5
+    install_requires=[
+        "typing"
```
Comment (guedou) [inline setup.py]: IMHO, this is a bad idea, and we should handle that yourself.

File: scapy/tools/UTscapy.py [hunk 10]
```diff
@@ -29,6 +29,13 @@
 from scapy.modules.six.moves import range
 from scapy.compat import base64_bytes, bytes_hex, plain_str
 
+ANNOTATIONS_MODE = False  # Debug: Set to True to use pyannotate
```
Comment (guedou) [inline scapy/tools/UTscapy.py]: Can it be a command line argument?

**Comment by guedou** [comment]
@gpotter2 the guide can be integrated into `CONTRIBUTING.md`

**Comment by guedou** [comment]
BTW, `http2.py` already uses `typing`.

File: .config/travis/install.sh [hunk 11]
```diff
@@ -2,13 +2,11 @@
 # Install on osx
 if [ "$TRAVIS_OS_NAME" = "osx" ]
 then
-  pip3 install tox
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply .config/travis/install.sh]: No but because I've removed the system exit 0 below, the rest of the code is executed.

File: .config/travis/install.sh [hunk 12]
```diff
@@ -26,7 +24,7 @@ then
 fi
 
 # Check pip
-sudo pip install --upgrade pip setuptools --ignore-installed
+python -m pip install --upgrade pip setuptools typing --ignore-installed
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply .config/travis/install.sh]: We call once `Python -c from Scapy.all import   conf; print(conf)`.

About pip and setuptools, tox doesn't update the ones it uses, so we need to do that manually.

File: scapy/__init__.py [hunk 13]
```diff
@@ -91,7 +100,7 @@ def _version():
 
 
 VERSION = __version__ = _version()
-VERSION_MAIN = re.search(r"[0-9.]+", VERSION).group()
+VERSION_MAIN = re.search(r"[0-9.]+", VERSION).group()  # type: ignore
```
[the remark this answers is not in the document]
Comment (gpotter2) [reply scapy/__init__.py]: Alright

File: scapy/main.py (hunk 8 above)
Comment (gpotter2) [reply scapy/main.py]: Good idea. It will cleaner and act as in http2.py

File: scapy/tools/UTscapy.py (hunk 10 above)
Comment (gpotter2) [reply scapy/tools/UTscapy.py]: Of course

**Comment by guedou** [comment]
BTW, can you remove line 2502 from scapy/contrib/http2.py ?

**Comment by gpotter2** [comment]
I think I've addressed all your points. I've also added `http2.py` to the list of files that get checked (with a few tweaks)

File: scapy/contrib/http2.py [hunk 14]
```diff
@@ -638,14 +636,13 @@ def _compute_value(self, pkt):
...
 
-
 if sys.version_info >= (3, 4):
     ABC = abc.ABC
 else:
     ABC = abc.ABCMeta('ABC', (), {})
 
 
-class HPackStringsInterface(ABC, Sized):
+class HPackStringsInterface(ABC, Sized):  # type: ignore
```
Comment (gpotter2) [inline scapy/contrib/http2.py]: I've used `# type: ignore` only in here, but that's very messy and hard to fix.
`abc.ABC` is already super tricky to get working (see the tweaks above)

File: .config/mypy/mypy.ini [hunk 15]
```diff
@@ -0,0 +1,10 @@
+[mypy]
+
+[mypy-IPython]
+ignore_missing_imports = True
+
+[mypy-colorama]
+ignore_missing_imports = True
+
+[mypy-traitlets.config.loader]
+ignore_missing_imports = True
```
Comment (guedou) [inline .config/mypy/mypy.ini]: Could you add the following lines? 
```
[mypy-scapy.modules.six,scapy.modules.winpcapy]
ignore_errors = True
```

They disable errors reported in modules that we will likely not bother to type.

Comment (gpotter2) [reply .config/mypy/mypy.ini]: Done

**Comment by gpotter2** [comment]
Should be ready to merge

**Comment by gpotter2** [comment]
Rebased

**Comment by p-l-** [comment]
Your call, @guedou.

**Comment by guedou** [comment]
@gpotter2 could you open an issue like https://github.com/secdev/scapy/issues/1277 to track the progress of adding types to Scapy?

**Comment by gpotter2** [comment]
This was supposed to be https://github.com/secdev/scapy/issues/2158.

**Comment by guedou** [comment]
Oh, I forgot about it ...
